File: gui/csvons_gui/lib/core/issue_filters.dart
import '../models/validation_report.dart';

enum IssueSortField {
  severity,
  file,
  rule,
  field,
  row,
  value,
  message,
}

List<ValidationIssue> filterAndSortIssues({
  required List<ValidationIssue> issues,
  String query = '',
  String severityFilter = 'all',
  String fileFilter = 'all',
  String ruleFilter = 'all',
  IssueSortField sortField = IssueSortField.severity,
  bool ascending = true,
}) {
  final normalizedQuery = query.trim().toLowerCase();
  final normalizedSeverityFilter = severityFilter.toLowerCase();
  final filtered = issues.where((issue) {
    final severityOk = normalizedSeverityFilter == 'all' ||
        issue.severity.toLowerCase() == normalizedSeverityFilter;
    if (!severityOk) return false;
    final fileOk = fileFilter == 'all' || issue.file == fileFilter;
    if (!fileOk) return false;
    final ruleOk = ruleFilter == 'all' || issue.rule == ruleFilter;
    if (!ruleOk) return false;
    if (normalizedQuery.isEmpty) return true;

    final blob =
        '${issue.message} ${issue.file} ${issue.rule} ${issue.field} ${issue.value ?? ''} ${issue.row ?? ''} ${issue.severity}'
            .toLowerCase();
    return blob.contains(normalizedQuery);
  }).toList(growable: false);

  int compare(ValidationIssue a, ValidationIssue b) {
    switch (sortField) {
      case IssueSortField.severity:
        final rankDelta = _severityRank(a.severity).compareTo(
          _severityRank(b.severity),
        );
        if (rankDelta != 0) {
          return rankDelta;
        }
        return a.severity.compareTo(b.severity);
      case IssueSortField.file:
        return a.file.compareTo(b.file);
      case IssueSortField.rule:
        return a.rule.compareTo(b.rule);
      case IssueSortField.field:
        return a.field.compareTo(b.field);
      case IssueSortField.row:
        return _rowSortValue(a.row).compareTo(_rowSortValue(b.row));
      case IssueSortField.value:
        return _valueSortValue(a.value).compareTo(_valueSortValue(b.value));
      case IssueSortField.message:
        return a.message.compareTo(b.message);
    }
  }

  filtered.sort((a, b) {
    var result = compare(a, b);
    if (result == 0) {
      result = _rowSortValue(a.row).compareTo(_rowSortValue(b.row));
    }
    if (result == 0) {
      result = a.message.compareTo(b.message);
    }
    return ascending ? result : -result;
  });
  return filtered;
}

int _rowSortValue(int? row) => row ?? 1 << 30;

String _valueSortValue(String? value) {
  if (value == null || value.isEmpty) {
    return '\uffff';
  }
  return value;
}

int _severityRank(String severity) {
  switch (severity.toLowerCase()) {
    case 'critical':
    case 'fatal':
      return 0;
    case 'error':
      return 1;
    case 'warning':
      return 2;
    case 'info':
      return 3;
    default:
      return 4;
  }
}

File: gui/electron/src/renderer/lib/issue-filters.ts
export type IssueSortField =
  | 'severity'
  | 'file'
  | 'rule'
  | 'field'
  | 'row'
  | 'value'
  | 'message';

export interface ValidationIssue {
  file?: string;
  rule?: string;
  field?: string;
  row?: number;
  value?: string;
  message: string;
  severity: string;
}

function severityRank(severity: string): number {
  switch (severity.toLowerCase()) {
    case 'critical':
      return 0;
    case 'error':
      return 1;
    case 'warning':
      return 2;
    case 'info':
      return 3;
    default:
      return 4;
  }
}

export function filterAndSortIssues(opts: {
  issues: ValidationIssue[];
  query?: string;
  severityFilter?: string;
  fileFilter?: string;
  ruleFilter?: string;
  sortField?: IssueSortField;
  ascending?: boolean;
}): ValidationIssue[] {
  const {
    issues,
    query = '',
    severityFilter = 'all',
    fileFilter = 'all',
    ruleFilter = 'all',
    sortField = 'severity',
    ascending = true,
  } = opts;

  let filtered = issues;

  if (severityFilter !== 'all') {
    filtered = filtered.filter(
      (i) => i.severity.toLowerCase() === severityFilter.toLowerCase(),
    );
  }

  if (fileFilter !== 'all') {
    filtered = filtered.filter((i) => i.file === fileFilter);
  }

  if (ruleFilter !== 'all') {
    filtered = filtered.filter((i) => i.rule === ruleFilter);
  }

  if (query.trim()) {
    const q = query.toLowerCase();
    filtered = filtered.filter((i) => {
      const blob = [
        i.message,
        i.file ?? '',
        i.rule ?? '',
        i.field ?? '',
        i.value ?? '',
        i.row != null ? String(i.row) : '',
        i.severity,
      ]
        .join(' ')
        .toLowerCase();
      return blob.includes(q);
    });
  }

  const sorted = [...filtered].sort((a, b) => {
    let cmp = 0;

    switch (sortField) {
      case 'severity':
        cmp = severityRank(a.severity) - severityRank(b.severity);
        if (cmp === 0) cmp = a.severity.localeCompare(b.severity);
        break;
      case 'row':
        cmp = (a.row ?? 1 << 30) - (b.row ?? 1 << 30);
        break;
      case 'file':
        cmp = (a.file ?? '').localeCompare(b.file ?? '');
        break;
      case 'rule':
        cmp = (a.rule ?? '').localeCompare(b.rule ?? '');
        break;
      case 'field':
        cmp = (a.field ?? '').localeCompare(b.field ?? '');
        break;
      case 'value':
        cmp = (a.value ?? '\uffff').localeCompare(b.value ?? '\uffff');
        break;
      case 'message':
        cmp = a.message.localeCompare(b.message);
        break;
    }

    if (cmp === 0) cmp = (a.row ?? 1 << 30) - (b.row ?? 1 << 30);
    if (cmp === 0) cmp = a.message.localeCompare(b.message);

    return ascending ? cmp : -cmp;
  });

  return sorted;
}
